### PR TITLE
MTHINC

### LINKED ISSUE #1200: MUSCL THINC right-state reconstruction uses already-overwritten left-state values
## Bug

In \`src/simulation/m_muscl.fpp\`, the THINC interface reconstruction computes the left-state first, overwriting the density ratio values. The subsequent right-state computation reads these already-overwritten values instead of the original ones.

**File:** \`src/simulation/m_muscl.fpp\`, lines 285-301

**Impact:** Right-state THINC reconstruction uses contaminated density ratios from the left-state computation. Currently masked by algebraic coincidence in simple cases, but produces wrong results when left and right densities differ significantly.

**Severity:** High — silent corruption of MUSCL-THINC interface reconstruction.

### LINKED ISSUE #1396: THINC/MTHINC: int_comp incompatible with viscosity and surface tension (reconstruction split silently disables compression)
## Summary

When `int_comp > 0` is combined with `viscous = T` or `surface_tension = T`, interface compression is **silently disabled** — no error, no warning, no sharpening.

## Root cause

In `m_rhs.fpp`, the reconstruction path splits into multiple `s_reconstruct_cell_boundary_values` calls with disjoint variable index ranges when viscosity or surface tension is enabled:

```
# Viscous case (Re_size > 0):
call s_reconstruct(vf(1 : cont%end), ...)       ! partial densities only
call s_reconstruct(vf(E   : sys_size), ...)      ! energy + adv (volume fractions)

# Surface tension case:
call s_reconstruct(vf(1   : E-1), ...)           ! cont + mom
call s_reconstruct_first_order(vf(E), ...)       ! energy
call s_reconstruct(vf(E+1 : sys_size), ...)      ! adv (volume fractions)
```

Inside `s_weno` and `s_muscl`, compression is guarded by:

```fortran
if (int_comp > 0 .and. v_size >= eqn_idx%adv%end) then
    call s_thinc_compression(...)
end if
```

`v_size = ubound(v_vf, 1)` for the current slice. In the split paths, no single slice contains both `cont` (needed for density ratios) and `adv` (volume fractions), so `v_size >= eqn_idx%adv%end` is never true and compression never fires.

## Current mitigation (merged in PR #1303)

`check_interface_compression` in `toolchain/mfc/case_validator.py` now prohibits these combinations explicitly:

```python
self.prohibit(int_comp != 0 and viscous,
    "int_comp > 0 is not supported with viscosity (reconstruction is split; compression would be silently skipped)")
self.prohibit(int_comp != 0 and surface_tension,
    "int_comp > 0 is not supported with surface tension (reconstruction is split; compression would be silently skipped)")
```

This turns a silent wrong result into a clear error at validation time.

## What full support would require

To actually support `int_comp` with viscosity or surface tension, the reconstruction in `m_rhs` would need to be restructured so that `cont` and `adv` variables are always reconstructed together in a single WENO/MUSCL call when `int_comp > 0`, allowing `s_thinc_compression` to access both sets of variables at once. This is non-trivial and warrants its own PR.

### PR DESCRIPTION
**Pull request by wilfonba**
## Description

This PR adds MTHINC to the code base and fixes the THINC bug in #1181. The existing golden files for THINC interface compression were preserved, and new golden files and tests were created for MTHINC.

### Type of change

- [x] New feature

### 3D advection case with and without MTHINC.

https://github.com/user-attachments/assets/7755a01b-77de-4896-9a77-84ffb44d4686

### 2D advection with `int_comp = 1` (THINC) 1 rank vs. 4 rank. Results are identical

https://github.com/user-attachments/assets/e2550e5d-19a9-40a7-a97c-e6eed6c38396

### 2D advection with `int_comp = 2` (MTHINC) 1 rank vs. 4 rank. Results are identical

https://github.com/user-attachments/assets/124d2509-fcd4-493a-86bd-402d3f2c0af7

### 3D advection with `int_comp = 2` (MTHINC) 1 rank vs. 8 rank. Results are identical
I didn't bother making a video for this one, but the output was identical when compared in ParaView.

### Performance tests in 3D
The following results are grind times for a 200^3 advection case run on 1 A100 with OpenACC. The overhead for MTHINC is ~3% while the overhead for THINC is ~1%. The grind times are:
(Updated 4/1/2026 after removing `parameter` from constants in `m_thinc.fpp`
- `int_comp = 0`: 0.87488               (No compression, baseline)
- `int_comp = 1`: 0.88406 (1%)   (THINC compression)
- `int_comp = 2`: 0.90199 (3%)  (MTHINC compression)

### `--nsys` results with `int_comp = 2`
This is the run time contribution for just the MTHINC routines.

Time (%)  Total Time (ns)  Instances    Avg (ns)      Med (ns)     Min (ns)    Max (ns)   StdDev (ns)                                           Name                                          
 --------  ---------------  ---------  ------------  ------------  ----------  ----------  -----------  ---------------------------------------------------------------------------------------
      1.3       75,807,250         75   1,010,763.3   1,028,360.0     183,969   1,159,626    118,287.9  m_thinc_s_compute_mthinc_normals_253_gpu 
      0.5       26,689,214         75     355,856.2     355,779.0     354,243     357,475        582.9  m_thinc_s_compute_mthinc_normals_228_gpu                                               
      0.3       19,284,544         75     257,127.3     257,666.0     229,922     259,746      3,433.0  m_thinc_s_thinc_compression_316_gpu__1                                                 
      0.3       19,228,227         75     256,376.4     256,931.0     229,282     258,531      3,376.6  m_thinc_s_thinc_compression_316_gpu                                                    
      0.3       19,204,612         75     256,061.5     256,354.0     229,890     258,786      3,361.1  m_thinc_s_thinc_compression_316_gpu__2 

## Checklist

- [x] I added or updated tests for new behavior
- [x] I updated documentation if user-facing behavior changed

See the [developer guide](https://mflowcode.github.io/documentation/contributing.html) for full coding standards.

<details>
<summary><strong>GPU changes</strong> (expand if you modified <code>src/simulation/</code>)</summary>

- [x] GPU results match CPU results
- [x] Tested on NVIDIA GPU or AMD GPU

</details>

## AI code reviews

Reviews are not triggered automatically. To request a review, comment on the PR:
- `@coderabbitai review` — incremental review (new changes only)
- `@coderabbitai full review` — full review from scratch
- `/review` — Qodo review
- `/improve` — Qodo code suggestions

### CODE REVIEW

**Comment by sbryngelson** [comment]
@wilfonba can you re-run a few sanity checks to make sure this works after i made a bunch of changes? then i guess we can mark it ready to review, which ensures it works on frontier/phoenix/whatever. then merge.

**Comment by wilfonba** [comment]
> @wilfonba can you re-run a few sanity checks to make sure this works after i made a bunch of changes? then i guess we can mark it ready to review, which ensures it works on frontier/phoenix/whatever. then merge.

Yep. I'll send you a message and open it as ready to review when I've taken care of that.

**Comment by sbryngelson** [comment]
PRs are ready to review (non-draft) when they are verified passing on amd gpu hardware + mpi (not the case here it seems). Setting from draft to ready to review triggers larger CI runs on every push.

**Comment by wilfonba** [comment]
@sbryngelson this is verified passing on NVIDIA with OpenACC/MP and AMD Flang with OpenMP now. It also gives correct results with parallelization. The existing PR comments from the draft state have been addressed as well.

**Comment by wilfonba** [comment]
Sheesh, seems like a weird edge case with compile-time parameters on CCE compilers is causing the issue with the tests.

**Comment by wilfonba** [comment]
Marking this ready for review again since I've run a bunch of manual tests since the most recent changes

File: src/simulation/m_thinc.fpp [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+                                        ! Right face (face_pos = +0.5)
+                                        aTHINC = f_mthinc_face_average(nh1, nh2, nh3, d_local, ic_beta, ${REC_DIR}$, 5e-1_wp, &
+                                                                       & num_dims)
+                                        if (aTHINC < ic_eps) aTHINC = ic_eps
+                                        if (aTHINC > 1._wp - ic_eps) aTHINC = 1._wp - ic_eps
+                                        vR_rs_vf_${XYZ}$ (j, k, l, contxb) = rho1*aTHINC
+                                        vR_rs_vf_${XYZ}$ (j, k, l, contxe) = rho2*(1._wp - aTHINC)
+                                        vR_rs_vf_${XYZ}$ (j, k, l, advxb) = aTHINC
+                                        vR_rs_vf_${XYZ}$ (j, k, l, advxe) = 1._wp - aTHINC
```
[the remark this answers is not in the document]
Comment (wilfonba) [reply src/simulation/m_thinc.fpp]: There is a case validation check for interface compression with > two fluids

**Comment by wilfonba** [comment]
@sbryngelson, I believe this is ready to merge. The failed benchmark shows good performance if you look at the generated comment. There was just some sort of error with the Slurm job. I can rerun it still if you want though.

**Comment by sbryngelson** [comment]
### Code review

Found 2 issues:

1. Missing `num_fluids == 2` constraint in `check_interface_compression()`. The THINC/MTHINC code in `s_thinc_compression` hardcodes updates to only `contxb`/`contxe` and `advxb`/`advxe`, and sets the second volume fraction to `1 - alpha_1`. This only works for exactly 2 fluids. With `num_fluids > 2`, intermediate partial densities and volume fractions are silently left unmodified, violating volume fraction normalization. The existing `surface_tension` constraint follows the correct pattern: `self.prohibit(surface_tension and num_fluids != 2, ...)`. (CLAUDE.md says "Every new parameter MUST be added in at least 3 places (4 if it has constraints)")

https://github.com/MFlowCode/MFC/blob/9c875b1c01c2d0542ad85321f4200cb3acffbacd/toolchain/mfc/case_validator.py#L369-L382

2. Init/finalize guard asymmetry for THINC module in `m_start_up.fpp`. `s_initialize_thinc_module()` is called inside `if (.not. igr .or. dummy)` (line 917), but `s_finalize_thinc_module()` is inside `else` of `if (igr)` (line 1085). When `igr=.true.` and `dummy=.true.`: init runs (allocates `mthinc_nhat`, `mthinc_d`), but finalize never runs because it is in the `else` branch which requires `igr=.false.`. This is a memory leak on the IGR dummy path. (CLAUDE.md says "Every `@:ALLOCATE(...)` MUST have a matching `@:DEALLOCATE(...)`")

https://github.com/MFlowCode/MFC/blob/9c875b1c01c2d0542ad85321f4200cb3acffbacd/src/simulation/m_start_up.fpp#L914-L926
https://github.com/MFlowCode/MFC/blob/9c875b1c01c2d0542ad85321f4200cb3acffbacd/src/simulation/m_start_up.fpp#L1085-L1096

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by wilfonba** [comment]
This was a good catch by Claude. A bunch of the tests for M/THINC were one fluid. I'll work on removing the existing garbage tests and creating new ones

**Review by sbryngelson — DISMISSED** [review]
Code looks good. Both issues from the April 5 review are fixed (num_fluids constraint and init/finalize guard). Merge conflicts resolved.

**Comment by sbryngelson** [comment]
## Claude Code Review

**PR #1303** — MTHINC

This PR adds MTHINC interface compression and refactors THINC into `m_thinc.fpp`. The Python-side validation is thorough and the THINC overwrite bug fix is correct. Three issues require attention before merge.

---

### Critical Issues

**1. GPU device memory never initialized for Gauss-quadrature constants — MTHINC produces garbage on GPU**
`src/simulation/m_thinc.fpp` lines 24–27

```fortran
real(wp) :: gq3_pts(3) = [...]
real(wp) :: gq3_wts(3) = [...]
real(wp) :: ln2 = 0.6931471805599453_wp
$:GPU_DECLARE(create='[gq3_pts, gq3_wts, ln2]')
```

`GPU_DECLARE(create=...)` allocates **uninitialized** device memory. It does NOT copy host values to the device. There is no `$:GPU_UPDATE(device='[gq3_pts, gq3_wts, ln2]')` call anywhere in the file — not in `s_initialize_thinc_module`, not in `s_thinc_compression`, nowhere.

These constants are read inside GPU `seq` device routines: `f_log_cosh` (uses `ln2`), `f_mthinc_volume_integral` and `f_mthinc_face_average` (use `gq3_pts`, `gq3_wts`). On GPU, all three read undefined values, silently producing wrong results for every MTHINC kernel invocation.

The codebase pattern is clear: `m_weno.fpp` declares arrays with `GPU_DECLARE(create=...)` and then calls `$:GPU_UPDATE(device=...)` inside `s_weno` before GPU loops. Fix — add to `s_initialize_thinc_module`:

```fortran
subroutine s_initialize_thinc_module()
    $:GPU_UPDATE(device='[gq3_pts, gq3_wts, ln2]')
    ...
end subroutine
```

---

### Important Issues

**2. MTHINC normals ignore grid spacing — incorrect on non-uniform and anisotropic grids**
`src/simulation/m_thinc.fpp` ~lines 165–174 (`s_compute_mthinc_normals`)

```fortran
nr_x = (v_vf(eqn_idx%adv%beg)%sf(j+1,k,l) - v_vf(eqn_idx%adv%beg)%sf(j-1,k,l))*5e-1_wp
nr_y = (v_vf(eqn_idx%adv%beg)%sf(j,k+1,l) - v_vf(eqn_idx%adv%beg)%sf(j,k-1,l))*5e-1_wp
```

No division by `dx`, `dy`, `dz`. The factor 0.5 assumes unit spacing in every direction. On non-uniform grids or whenever `dx ≠ dy` (common in 2D/3D cases), the computed normal direction is distorted by the grid aspect ratio and does not represent the physical interface orientation, degrading MTHINC to worse-than-1D behavior.

Fix:
```fortran
nr_x = (alpha_jp1 - alpha_jm1) / (x_cc(j+1) - x_cc(j-1))
nr_y = (alpha_kp1 - alpha_km1) / (y_cc(k+1) - y_cc(k-1))
nr_z = (alpha_lp1 - alpha_lm1) / (z_cc(l+1) - z_cc(l-1))
```

**3. 4-space indentation throughout `m_thinc.fpp` — project standard is 2-space**

Project standard enforced by `./mfc.sh format`. Run `./mfc.sh format -j 8` before merge.

---

### Strengths

- Python-side validation in `case_validator.py` is comprehensive: enforces `num_fluids=2`, `n>0` for MTHINC, and WENO/MUSCL order ≥ 3 requirements.
- `GPU_DECLARE(create=...)` + `GPU_UPDATE(device=...)` for `int_comp`, `ic_eps`, `ic_beta` in `m_global_parameters.fpp` follows the correct pattern.
- The THINC overwrite bug fix is correct: `rho_b`/`rho_e` are read from unmodified `v_rs_ws` before any writes to `vL`/`vR`.
- `@:ALLOCATE`/`@:DEALLOCATE` pairing for `mthinc_nhat` and `mthinc_d` is correct and symmetric.
- MPI: `int_comp` correctly moved from the logical broadcast list to the integer list in `m_mpi_proxy.fpp`.
- `GPU_ROUTINE(parallelism='[seq]')` correctly applied to all device functions.

**Comment by sbryngelson** [comment]
Reviewing the two open Qodo comments from 2026-05-04:

**1. Normal scaling on non-uniform grids (`m_thinc.fpp:259`)**

After reading the code closely, this is not a bug in the current formulation. The MTHINC implementation works entirely in reference space — the unit cell ξ ∈ [−½, ½]^ndim. `f_mthinc_volume_integral` and `f_mthinc_face_average` both integrate over reference coordinates with face positions fixed at ±½, so the normal vector must also be in reference space. The code computes `(α(j+1) − α(j−1)) * 0.5`, which is the reference-space directional derivative up to a normalization constant — correct. Qodo's suggested fix (divide by dx/dy/dz) would give the physical-space gradient, which is inconsistent with the rest of the reference-space formulation.

There is a real accuracy limitation on strongly non-uniform (stretched) grids, where the physical and reference-space interface normals diverge. But this is a formulation-level constraint, not a code bug, and it does not affect the primary use case (uniform grids). Tracked in #1305 for future work.

**2. Compression silently skipped with viscosity or surface tension (`m_weno.fpp:1395`)**

This is a genuine issue. When `viscous = T` or `surface_tension = T`, `m_rhs` splits reconstruction into multiple calls with disjoint variable index ranges — one for `cont`, one for `E` through `sys_size`. The guard `v_size >= eqn_idx%adv%end` prevents out-of-bounds access but as a side effect silently disables interface compression in every call, because `cont` and `adv` are never present in the same slice. There is no existing validator prohibition for these combinations.

Fixed in this branch by adding prohibitions in `check_interface_compression` so that `int_comp > 0` with `viscous = T` or `surface_tension = T` now fails at validation time with a clear message rather than silently producing unsharpened results. Full support for those combinations (restructuring reconstruction to keep `cont`+`adv` together when `int_comp > 0`) is tracked in #1306.

**Comment by sbryngelson** [comment]
Correction to my previous comment: the issue numbers are #1395 (MTHINC normals on non-uniform grids) and #1396 (int_comp + viscosity/surface tension), not #1305/#1306.

**Comment by wilfonba** [comment]
These are good findings. Do these need to be resolved in this PR, or can they be added as prohibited conditions and implemented separately?

**Comment by sbryngelson** [comment]
## Automated Code Review

**Summary:** 3 critical issues, 3 important issues.

### Critical Issues

**1. Missing `use m_nvtx` in `m_muscl.fpp`** (confidence: 100%)

`s_muscl` calls `nvtxStartRange("WENO-INTCOMP")` and `nvtxEndRange()` but `m_muscl.fpp` has no `use m_nvtx` in its module-level use list. `m_weno.fpp` correctly includes it for its own NVTX calls. This will fail to compile on CPU-only or non-NVIDIA builds.

**Fix:** Add `use m_nvtx` to the module-level use statements in `m_muscl.fpp`.

---

**2. MTHINC normal gradients not scaled by grid spacing — incorrect normals on non-square grids** (confidence: 85%)

In `s_compute_mthinc_normals`:
```fortran
nr_x = (alpha(j+1) - alpha(j-1)) * 5e-1_wp
nr_y = (alpha(j, k+1) - alpha(j, k-1)) * 5e-1_wp
```
These are unscaled finite differences (Δα/2), not physical gradients (Δα/(2Δx)). Although the normal is subsequently normalized to unit length, the *direction* is wrong when Δx ≠ Δy ≠ Δz. On stretched or non-cubic grids the normal is biased toward the coarser-spaced direction.

**Fix:** Divide `nr_x` by the local cell width (`x_cb(j) - x_cb(j-1)`), `nr_y` by the local cell height, etc.

---

**3. `muscl_order == 1` + `int_comp > 0` prohibition present in `m_checker.fpp` and `case_validator.py` ✓, but no Fortran-side check for `weno_order == 1` + `int_comp != 0`** (confidence: 95%)

`case_validator.py` correctly prohibits `weno_order == 1 and int_comp != 0`, but this check is absent from `m_checker.fpp`'s WENO validation section. A user supplying a hand-written `.inp` file bypasses Python validation and hits undefined behavior.

---

### Important Issues

**4. `f_log_cosh_diff` atanh clamping with `epsilon(1._wp)` may underestimate in mixed-precision builds** (confidence: 82%)

```fortran
res = 2._wp*atanh(sign(min(abs(t), 1._wp - epsilon(1._wp)), t))
```
In double precision `epsilon(1._wp) ≈ 2.2e-16`, which is fine — `atanh(1 - ε)` returns ~36.7. But in mixed-precision (`wp = real16`), `epsilon(1._wp) ≈ 0.001`, so `atanh` is only called up to argument 0.999 and the result underestimates the true value by up to ~3.5 on sharp interfaces. Consider using a tighter absolute minimum clamp (e.g., `1._wp - 1.e-7_wp`) rather than `epsilon(1._wp)`.

**5. `id_norm` derived-type passed via `copyin` in GPU_PARALLEL_LOOP** (confidence: 82%)

`copyin='[id_norm]'` on a local stack `type(int_bounds_info), dimension(3)` variable is well-defined for OpenACC/NVIDIA but implementation-defined for OpenMP target offload (AMD flang, Cray ftn). This may silently produce incorrect loop bounds on non-NVIDIA GPU builds. The safer pattern is to extract the needed integer scalars and pass them as `firstprivate`.

**6. Missing Fortran-side `@:PROHIBIT` for `int_comp == 2` with `n == 0`** (confidence: 85%)

`case_validator.py` correctly prohibits MTHINC in 1D, but `m_checker.fpp` has no equivalent check. Runtime behavior is actually safe (graceful fallback to THINC-1D path), but the missing Fortran guard means `.inp`-based runs degrade silently rather than aborting with a clear message.

### Strengths

- The THINC bug fix from #1181 is correctly implemented: `rho_b`/`rho_e` are computed from the original workspace before any face values are written
- `f_log_cosh` and `f_thinc_integral_1d` have appropriate large-argument branch cutoffs
- `@:ALLOCATE`/`@:DEALLOCATE` pairing in thinc module is correct and guarded by `int_comp == 2`
- All 4 parameter-system locations updated for `int_comp`, `ic_eps`, `ic_beta`
- `muscl_order == 1` + `int_comp > 0` prohibition present in both `m_checker.fpp` and `case_validator.py`
- MPI broadcasts for `int_comp`, `ic_eps`, `ic_beta` present in `m_mpi_proxy.fpp`
